### DISCUSSION TITLE
Assignment solutions with Jupyter Book

### DISCUSSION THREAD

**Discussion by fmaussion** [executablebooks/meta#177] [Q&A]
**TL;DR**: jump to the end of the post.

I provide non graded, weekly assignments to students. At the end of the week, I upload the assignment solutions to the book website. A typical situation looks like:

Week 01
- Assignment_01.ipynb
- Assignment_01_solution.ipynb

Week 02
- Assignment_02.ipynb
- Assignment_02_solution.ipynb

(example: https://fabienmaussion.info/climate_system/ )

It does what I want: I can easily handle the timing of uploading the assignment and uploading their solution by adding the solution to the toc file, and it looks OK. Latecomers can still download both the empty assignment (if they want to work on it) and their solution (if they are lazy).

The main problem with this design is that I have to maintain two nearly identical notebooks (one with the solution and one with the assignment).

Is there a way to do this better?

I'm aware of https://github.com/executablebooks/sphinx-exercise which is nice but not adapted to this situation (where an entire notebook is the assignment) and nbgrader which is way too overkill for what I want. 

The closest setup to what I'd like to have is the hide option in jupyter book (https://jupyterbook.org/interactive/hiding.html) but this still won't make it because I'd need to manually edit all the cells tags AND since the students download the notebooks from the website they'll see the solutions.

Now that I've written all this I think I know what I'd need. I need a tool that converts a solution notebook (with solutions cells tagged as such) to a notebook where the solution cells are removed. Does someone know if such a tool exists?

**Comment by fmaussion** [comment]
OK, I found this: https://nbconvert.readthedocs.io/en/latest/removing_cells.html 

I think that with a relatively simple script and config file this is going to do what I want

**Reply by choldgraf** [reply]
I made a little lightweight tool that accomplished something like this when I was teaching courses and had the same problem. Maybe you'd find it useful? https://github.com/choldgraf/nbclean

**Reply by fmaussion** [reply]
Yes!!! Awesome. Such a great community :unicorn:

**Reply by choldgraf** [reply]
note that I haven't updated that in a while, though PRs and suggestions are definitely welcome!

**Reply by fmaussion** [reply]
I'll have a more detailed look tomorrow, but it looks like a lot of `nbclean`'s functionality is now available in `nbconvert`'s  "Removing cells" functionality?  https://nbconvert.readthedocs.io/en/latest/removing_cells.html

**Reply by choldgraf** [reply]
Actually it is the reverse haha - I made nbclean because I wanted "remove cells" functionality from nbgrader but without all of the extra baggage that came with nbgrader. So nbclean can be thought of as an attempt at pulling out specific functionality from nbgrader and modularizing it in a single package
